File: src/bmsx/render/backend/software/vdp_framebuffer_rasterizer.ts
import {
	type Layer2D,
} from '../../../machine/devices/vdp/contracts';
import {
	VdpBlitterCommandBuffer,
	type VdpBlitterSource,
	VDP_BLITTER_IMPLICIT_CLEAR,
	VDP_BLITTER_OPCODE_BATCH_BLIT,
	VDP_BLITTER_OPCODE_BLIT,
	VDP_BLITTER_OPCODE_CLEAR,
	VDP_BLITTER_OPCODE_DRAW_LINE,
	VDP_BLITTER_OPCODE_FILL_RECT,
} from '../../../machine/devices/vdp/blitter';
import type { VDP } from '../../../machine/devices/vdp/vdp';

export class VdpFrameBufferRasterizer {
	private frameBufferPriorityLayer = new Uint8Array(0);
	private frameBufferPriorityZ = new Float32Array(0);
	private frameBufferPrioritySeq = new Uint32Array(0);
	private readonly latchedSourceScratch: VdpBlitterSource = { surfaceId: 0, srcX: 0, srcY: 0, width: 0, height: 0 };

	public constructor(private readonly vdp: VDP) {}

	public executeFrameBufferCommands(commands: VdpBlitterCommandBuffer, frameWidth: number, frameHeight: number, pixels: Uint8Array): void {
		if (commands.length === 0) {
			return;
		}
		this.resizeFrameBufferPriorityStorage(frameWidth * frameHeight);
		if (commands.opcode[0] !== VDP_BLITTER_OPCODE_CLEAR) {
			this.fillFrameBuffer(pixels, VDP_BLITTER_IMPLICIT_CLEAR);
		}
		this.resetFrameBufferPriority();
		for (let index = 0; index < commands.length; index += 1) {
			const opcode = commands.opcode[index];
			if (opcode === VDP_BLITTER_OPCODE_CLEAR) {
				this.fillFrameBuffer(pixels, commands.color[index]);
				this.resetFrameBufferPriority();
				continue;
			}
			const layer = commands.layer[index] as Layer2D;
			const priority = commands.priority[index];
			const sequence = commands.seq[index];
			const color = commands.color[index];
			if (opcode === VDP_BLITTER_OPCODE_FILL_RECT) {
				this.rasterizeFrameBufferFill(pixels, frameWidth, frameHeight, commands.x0[index], commands.y0[index], commands.x1[index], commands.y1[index], color, layer, priority, sequence);
				continue;
			}
			if (opcode === VDP_BLITTER_OPCODE_DRAW_LINE) {
				this.rasterizeFrameBufferLine(pixels, frameWidth, frameHeight, commands.x0[index], commands.y0[index], commands.x1[index], commands.y1[index], commands.thickness[index], color, layer, priority, sequence);
				continue;
			}
			if (opcode === VDP_BLITTER_OPCODE_BLIT) {
				const source = this.latchedSourceScratch;
				source.surfaceId = commands.sourceSurfaceId[index];
				source.srcX = commands.sourceSrcX[index];
				source.srcY = commands.sourceSrcY[index];
				source.width = commands.sourceWidth[index];
				source.height = commands.sourceHeight[index];
				this.rasterizeFrameBufferBlit(pixels, frameWidth, frameHeight, source, commands.dstX[index], commands.dstY[index], commands.width[index], commands.height[index], commands.flipH[index] !== 0, commands.flipV[index] !== 0, color, layer, priority, sequence);
				continue;
			}
			if (opcode === VDP_BLITTER_OPCODE_BATCH_BLIT) {
				const firstItem = commands.batchBlitFirstEntry[index];
				const itemEnd = firstItem + commands.batchBlitItemCount[index];
				if (commands.hasBackgroundColor[index] !== 0) {
					for (let itemIndex = firstItem; itemIndex < itemEnd; itemIndex += 1) {
						this.rasterizeFrameBufferFill(pixels, frameWidth, frameHeight, commands.batchBlitDstX[itemIndex], commands.batchBlitDstY[itemIndex], commands.batchBlitDstX[itemIndex] + commands.batchBlitAdvance[itemIndex], commands.batchBlitDstY[itemIndex] + commands.lineHeight[index], commands.backgroundColor[index], layer, priority, sequence);
					}
				}
				for (let itemIndex = firstItem; itemIndex < itemEnd; itemIndex += 1) {
					const source = this.latchedSourceScratch;
					source.surfaceId = commands.batchBlitSurfaceId[itemIndex];
					source.srcX = commands.batchBlitSrcX[itemIndex];
					source.srcY = commands.batchBlitSrcY[itemIndex];
					source.width = commands.batchBlitWidth[itemIndex];
					source.height = commands.batchBlitHeight[itemIndex];
					this.rasterizeFrameBufferBlit(pixels, frameWidth, frameHeight, source, commands.batchBlitDstX[itemIndex], commands.batchBlitDstY[itemIndex], 1, 1, false, false, color, layer, priority, sequence);
				}
				continue;
			}
		}
	}

	private resizeFrameBufferPriorityStorage(pixelCount: number): void {
		if (this.frameBufferPriorityLayer.length === pixelCount) {
			return;
		}
		this.frameBufferPriorityLayer = new Uint8Array(pixelCount);
		this.frameBufferPriorityZ = new Float32Array(pixelCount);
		this.frameBufferPrioritySeq = new Uint32Array(pixelCount);
	}

	private resetFrameBufferPriority(): void {
		this.frameBufferPriorityLayer.fill(0);
		this.frameBufferPriorityZ.fill(Number.NEGATIVE_INFINITY);
		this.frameBufferPrioritySeq.fill(0);
	}

	private fillFrameBuffer(pixels: Uint8Array, color: number): void {
		const r = (color >>> 16) & 0xff;
		const g = (color >>> 8) & 0xff;
		const b = color & 0xff;
		const a = (color >>> 24) & 0xff;
		for (let pixelIndex = 0; pixelIndex < pixels.length; pixelIndex += 4) {
			pixels[pixelIndex + 0] = r;
			pixels[pixelIndex + 1] = g;
			pixels[pixelIndex + 2] = b;
			pixels[pixelIndex + 3] = a;
		}
	}

	private blendFrameBufferPixel(pixels: Uint8Array, index: number, r: number, g: number, b: number, a: number, layer: Layer2D, priority: number, seq: number): void {
		if (a === 0) {
			return;
		}
		const pixelIndex = index / 4;
		const currentLayer = this.frameBufferPriorityLayer[pixelIndex] as Layer2D;
		if (layer < currentLayer) {
			return;
		}
		if (layer === currentLayer) {
			const currentPriority = this.frameBufferPriorityZ[pixelIndex];
			if (priority < currentPriority) {
				return;
			}
			if (priority === currentPriority && seq < this.frameBufferPrioritySeq[pixelIndex]) {
				return;
			}
		}
		if (a === 255) {
			pixels[index + 0] = r;
			pixels[index + 1] = g;
			pixels[index + 2] = b;
			pixels[index + 3] = 255;
			this.frameBufferPriorityLayer[pixelIndex] = layer;
			this.frameBufferPriorityZ[pixelIndex] = priority;
			this.frameBufferPrioritySeq[pixelIndex] = seq;
			return;
		}
		const inverse = 255 - a;
		pixels[index + 0] = ((r * a) + (pixels[index + 0] * inverse) + 127) / 255;
		pixels[index + 1] = ((g * a) + (pixels[index + 1] * inverse) + 127) / 255;
		pixels[index + 2] = ((b * a) + (pixels[index + 2] * inverse) + 127) / 255;
		pixels[index + 3] = a + ((pixels[index + 3] * inverse) + 127) / 255;
		this.frameBufferPriorityLayer[pixelIndex] = layer;
		this.frameBufferPriorityZ[pixelIndex] = priority;
		this.frameBufferPrioritySeq[pixelIndex] = seq;
	}

	private rasterizeFrameBufferFill(pixels: Uint8Array, frameWidth: number, frameHeight: number, x0: number, y0: number, x1: number, y1: number, color: number, layer: Layer2D, priority: number, seq: number): void {
		const r = (color >>> 16) & 0xff;
		const g = (color >>> 8) & 0xff;
		const b = color & 0xff;
		const a = (color >>> 24) & 0xff;
		let left = x0;
		let top = y0;
		let right = x1;
		let bottom = y1;
		if (right < left) {
			const swap = left;
			left = right;
			right = swap;
		}
		if (bottom < top) {
			const swap = top;
			top = bottom;
			bottom = swap;
		}
		if (left < 0) left = 0;
		if (top < 0) top = 0;
		if (right > frameWidth) right = frameWidth;
		if (bottom > frameHeight) bottom = frameHeight;
		for (let y = top; y < bottom; y += 1) {
			let index = (y * frameWidth + left) * 4;
			for (let x = left; x < right; x += 1) {
				this.blendFrameBufferPixel(pixels, index, r, g, b, a, layer, priority, seq);
				index += 4;
			}
		}
	}

	private rasterizeFrameBufferLine(pixels: Uint8Array, frameWidth: number, frameHeight: number, x0: number, y0: number, x1: number, y1: number, thicknessValue: number, color: number, layer: Layer2D, priority: number, seq: number): void {
		const r = (color >>> 16) & 0xff;
		const g = (color >>> 8) & 0xff;
		const b = color & 0xff;
		const a = (color >>> 24) & 0xff;
		let currentX = x0;
		let currentY = y0;
		const targetX = x1;
		const targetY = y1;
		const dx = Math.abs(targetX - currentX);
		const dy = Math.abs(targetY - currentY);
		const sx = currentX < targetX ? 1 : -1;
		const sy = currentY < targetY ? 1 : -1;
		let err = dx - dy;
		const thickness = thicknessValue;
		while (true) {
			const half = thickness >> 1;
			for (let yy = currentY - half; yy < currentY - half + thickness; yy += 1) {
				if (yy < 0 || yy >= frameHeight) {
					continue;
				}
				for (let xx = currentX - half; xx < currentX - half + thickness; xx += 1) {
					if (xx < 0 || xx >= frameWidth) {
						continue;
					}
					this.blendFrameBufferPixel(pixels, (yy * frameWidth + xx) * 4, r, g, b, a, layer, priority, seq);
				}
			}
			if (currentX === targetX && currentY === targetY) {
				return;
			}
			const e2 = err << 1;
			if (e2 > -dy) {
				err -= dy;
				currentX += sx;
			}
			if (e2 < dx) {
				err += dx;
				currentY += sy;
			}
		}
	}

	private rasterizeFrameBufferBlit(pixels: Uint8Array, frameWidth: number, frameHeight: number, source: VdpBlitterSource, dstX: number, dstY: number, dstW: number, dstH: number, flipH: boolean, flipV: boolean, color: number, layer: Layer2D, priority: number, seq: number): void {
		const colorR = (color >>> 16) & 0xff;
		const colorG = (color >>> 8) & 0xff;
		const colorB = color & 0xff;
		const colorA = (color >>> 24) & 0xff;
		const sourceSlot = this.vdp.resolveFrameBufferExecutionSource(source.surfaceId);
		if (sourceSlot === null) {
			return;
		}
		const sourcePixels = sourceSlot.cpuReadback;
		const sourceStride = sourceSlot.surfaceWidth * 4;
		let srcY = 0;
		let srcYRemainder = 0;
		for (let y = 0; y < dstH; y += 1) {
			const targetY = dstY + y;
			if (targetY >= 0 && targetY < frameHeight) {
				const sampleSourceY = flipV ? source.height - 1 - srcY : srcY;
				let srcX = 0;
				let srcXRemainder = 0;
				for (let x = 0; x < dstW; x += 1) {
					const targetX = dstX + x;
					if (targetX >= 0 && targetX < frameWidth) {
						const sampleSourceX = flipH ? source.width - 1 - srcX : srcX;
						const sampleX = source.srcX + sampleSourceX;
						const sampleY = source.srcY + sampleSourceY;
						if (sampleX < sourceSlot.surfaceWidth && sampleY < sourceSlot.surfaceHeight) {
							const srcIndex = sampleY * sourceStride + sampleX * 4;
							const srcA = sourcePixels[srcIndex + 3];
							if (srcA !== 0) {
								const outA = (srcA * colorA + 127) / 255;
								const outR = (sourcePixels[srcIndex + 0] * colorR + 127) / 255;
								const outG = (sourcePixels[srcIndex + 1] * colorG + 127) / 255;
								const outB = (sourcePixels[srcIndex + 2] * colorB + 127) / 255;
								this.blendFrameBufferPixel(pixels, (targetY * frameWidth + targetX) * 4, outR, outG, outB, outA, layer, priority, seq);
							}
						}
					}
					srcXRemainder += source.width;
					while (srcXRemainder >= dstW) {
						srcX += 1;
						srcXRemainder -= dstW;
					}
				}
			}
			srcYRemainder += source.height;
			while (srcYRemainder >= dstH) {
				srcY += 1;
				srcYRemainder -= dstH;
			}
		}
	}

}

File: src/bmsx_cpp/render/backend/software/vdp_framebuffer_rasterizer.cpp
#include "render/backend/software/vdp_framebuffer_rasterizer.h"

#include "machine/devices/vdp/vdp.h"
#include <algorithm>
#include <limits>
#include <utility>

namespace bmsx {
namespace {

constexpr VdpFrameBufferColor VDP_BLITTER_IMPLICIT_CLEAR_COLOR{0u, 0u, 0u, 255u};

} // namespace

VdpFrameBufferRasterizer::VdpFrameBufferRasterizer(VDP& vdp)
	: m_vdp(vdp) {}

void VdpFrameBufferRasterizer::executeFrameBufferCommands(const VdpBlitterCommand& commands, u32 frameWidth, u32 frameHeight, std::vector<u8>& pixels) {
	if (commands.length == 0u) {
		return;
	}
	resizeFrameBufferPriorityStorage(static_cast<size_t>(frameWidth) * static_cast<size_t>(frameHeight));
	if (commands.opcode[0] != VdpBlitterCommandType::Clear) {
		fillFrameBuffer(pixels, VDP_BLITTER_IMPLICIT_CLEAR_COLOR);
	}
	resetFrameBufferPriority();
	for (size_t index = 0u; index < commands.length; ++index) {
		const VdpBlitterCommandType opcode = commands.opcode[index];
		if (opcode == VdpBlitterCommandType::Clear) {
			fillFrameBuffer(pixels, unpackArgbColor(commands.color[index]));
			resetFrameBufferPriority();
			continue;
		}
		const Layer2D layer = commands.layer[index];
		const f32 priority = commands.priority[index];
		const u32 sequence = commands.seq[index];
		const VdpFrameBufferColor color = unpackArgbColor(commands.color[index]);
		switch (opcode) {
			case VdpBlitterCommandType::FillRect:
				rasterizeFrameBufferFill(pixels, frameWidth, frameHeight, commands.x0[index], commands.y0[index], commands.x1[index], commands.y1[index], color, layer, priority, sequence);
				break;
			case VdpBlitterCommandType::DrawLine:
				rasterizeFrameBufferLine(pixels, frameWidth, frameHeight, commands.x0[index], commands.y0[index], commands.x1[index], commands.y1[index], commands.thickness[index], color, layer, priority, sequence);
				break;
			case VdpBlitterCommandType::Blit:
				m_latchedSourceScratch.surfaceId = commands.sourceSurfaceId[index];
				m_latchedSourceScratch.srcX = commands.sourceSrcX[index];
				m_latchedSourceScratch.srcY = commands.sourceSrcY[index];
				m_latchedSourceScratch.width = commands.sourceWidth[index];
				m_latchedSourceScratch.height = commands.sourceHeight[index];
				rasterizeFrameBufferBlit(pixels, frameWidth, frameHeight, m_latchedSourceScratch, commands.dstX[index], commands.dstY[index], commands.width[index], commands.height[index], commands.flipH[index] != 0u, commands.flipV[index] != 0u, color, layer, priority, sequence);
				break;
			case VdpBlitterCommandType::BatchBlit: {
				const size_t firstItem = commands.batchBlitFirstEntry[index];
				const size_t itemEnd = firstItem + commands.batchBlitItemCount[index];
				if (commands.hasBackgroundColor[index] != 0u) {
					const VdpFrameBufferColor background = unpackArgbColor(commands.backgroundColor[index]);
					for (size_t itemIndex = firstItem; itemIndex < itemEnd; ++itemIndex) {
						rasterizeFrameBufferFill(
							pixels,
							frameWidth,
							frameHeight,
							commands.batchBlitDstX[itemIndex],
							commands.batchBlitDstY[itemIndex],
							commands.batchBlitDstX[itemIndex] + static_cast<i32>(commands.batchBlitAdvance[itemIndex]),
							commands.batchBlitDstY[itemIndex] + static_cast<i32>(commands.lineHeight[index]),
							background,
							layer,
							priority,
							sequence
						);
					}
				}
				for (size_t itemIndex = firstItem; itemIndex < itemEnd; ++itemIndex) {
					m_latchedSourceScratch.surfaceId = commands.batchBlitSurfaceId[itemIndex];
					m_latchedSourceScratch.srcX = commands.batchBlitSrcX[itemIndex];
					m_latchedSourceScratch.srcY = commands.batchBlitSrcY[itemIndex];
					m_latchedSourceScratch.width = commands.batchBlitWidth[itemIndex];
					m_latchedSourceScratch.height = commands.batchBlitHeight[itemIndex];
					rasterizeFrameBufferBlit(pixels, frameWidth, frameHeight, m_latchedSourceScratch, commands.batchBlitDstX[itemIndex], commands.batchBlitDstY[itemIndex], static_cast<i32>(commands.batchBlitWidth[itemIndex]), static_cast<i32>(commands.batchBlitHeight[itemIndex]), false, false, color, layer, priority, sequence);
				}
				break;
			}
			case VdpBlitterCommandType::Clear:
				break;
		}
	}
}

void VdpFrameBufferRasterizer::resizeFrameBufferPriorityStorage(size_t pixelCount) {
	if (m_frameBufferPriorityLayer.size() == pixelCount) {
		return;
	}
	m_frameBufferPriorityLayer.resize(pixelCount);
	m_frameBufferPriorityZ.resize(pixelCount);
	m_frameBufferPrioritySeq.resize(pixelCount);
}

void VdpFrameBufferRasterizer::resetFrameBufferPriority() {
	std::fill(m_frameBufferPriorityLayer.begin(), m_frameBufferPriorityLayer.end(), static_cast<u8>(Layer2D::World));
	std::fill(m_frameBufferPriorityZ.begin(), m_frameBufferPriorityZ.end(), -std::numeric_limits<f32>::infinity());
	std::fill(m_frameBufferPrioritySeq.begin(), m_frameBufferPrioritySeq.end(), 0u);
}

void VdpFrameBufferRasterizer::fillFrameBuffer(std::vector<u8>& pixels, const VdpFrameBufferColor& color) {
	for (size_t index = 0u; index < pixels.size(); index += 4u) {
		pixels[index + 0u] = color.r;
		pixels[index + 1u] = color.g;
		pixels[index + 2u] = color.b;
		pixels[index + 3u] = color.a;
	}
}

void VdpFrameBufferRasterizer::blendFrameBufferPixel(std::vector<u8>& pixels, size_t pixelIndex, u8 r, u8 g, u8 b, u8 a, Layer2D layer, f32 priority, u32 seq) {
	if (a == 0u) {
		return;
	}
	const auto currentLayer = static_cast<Layer2D>(m_frameBufferPriorityLayer[pixelIndex]);
	if (layer < currentLayer) {
		return;
	}
	if (layer == currentLayer) {
		const f32 currentPriority = m_frameBufferPriorityZ[pixelIndex];
		if (priority < currentPriority) {
			return;
		}
		if (priority == currentPriority && seq < m_frameBufferPrioritySeq[pixelIndex]) {
			return;
		}
	}
	const size_t byteIndex = pixelIndex * 4u;
	if (a == 255u) {
		pixels[byteIndex + 0u] = r;
		pixels[byteIndex + 1u] = g;
		pixels[byteIndex + 2u] = b;
		pixels[byteIndex + 3u] = 255u;
		m_frameBufferPriorityLayer[pixelIndex] = static_cast<u8>(layer);
		m_frameBufferPriorityZ[pixelIndex] = priority;
		m_frameBufferPrioritySeq[pixelIndex] = seq;
		return;
	}
	const u32 inverse = 255u - a;
	const u32 dstR = pixels[byteIndex + 0u];
	const u32 dstG = pixels[byteIndex + 1u];
	const u32 dstB = pixels[byteIndex + 2u];
	const u32 dstA = pixels[byteIndex + 3u];
	pixels[byteIndex + 0u] = static_cast<u8>(((static_cast<u32>(r) * a) + (dstR * inverse) + 127u) / 255u);
	pixels[byteIndex + 1u] = static_cast<u8>(((static_cast<u32>(g) * a) + (dstG * inverse) + 127u) / 255u);
	pixels[byteIndex + 2u] = static_cast<u8>(((static_cast<u32>(b) * a) + (dstB * inverse) + 127u) / 255u);
	pixels[byteIndex + 3u] = static_cast<u8>(a + ((dstA * inverse) + 127u) / 255u);
	m_frameBufferPriorityLayer[pixelIndex] = static_cast<u8>(layer);
	m_frameBufferPriorityZ[pixelIndex] = priority;
	m_frameBufferPrioritySeq[pixelIndex] = seq;
}

void VdpFrameBufferRasterizer::rasterizeFrameBufferFill(std::vector<u8>& pixels, u32 frameWidth, u32 frameHeight, i32 x0, i32 y0, i32 x1, i32 y1, const VdpFrameBufferColor& color, Layer2D layer, f32 priority, u32 seq) {
	const i32 frameBufferWidth = static_cast<i32>(frameWidth);
	const i32 frameBufferHeight = static_cast<i32>(frameHeight);
	i32 left = x0;
	i32 top = y0;
	i32 right = x1;
	i32 bottom = y1;
	if (right < left) {
		std::swap(left, right);
	}
	if (bottom < top) {
		std::swap(top, bottom);
	}
	left = std::max(0, left);
	top = std::max(0, top);
	right = std::min(frameBufferWidth, right);
	bottom = std::min(frameBufferHeight, bottom);
	for (i32 y = top; y < bottom; ++y) {
		size_t index = static_cast<size_t>(y) * static_cast<size_t>(frameBufferWidth) + static_cast<size_t>(left);
		for (i32 x = left; x < right; ++x) {
			blendFrameBufferPixel(pixels, index, color.r, color.g, color.b, color.a, layer, priority, seq);
			index += 1u;
		}
	}
}

void VdpFrameBufferRasterizer::rasterizeFrameBufferLine(std::vector<u8>& pixels, u32 frameWidth, u32 frameHeight, i32 x0, i32 y0, i32 x1, i32 y1, i32 thicknessValue, const VdpFrameBufferColor& color, Layer2D layer, f32 priority, u32 seq) {
	const i32 frameBufferWidth = static_cast<i32>(frameWidth);
	const i32 frameBufferHeight = static_cast<i32>(frameHeight);
	i32 currentX = x0;
	i32 currentY = y0;
	const i32 targetX = x1;
	const i32 targetY = y1;
	const i32 dx = targetX >= currentX ? targetX - currentX : currentX - targetX;
	const i32 dy = targetY >= currentY ? targetY - currentY : currentY - targetY;
	const i32 sx = currentX < targetX ? 1 : -1;
	const i32 sy = currentY < targetY ? 1 : -1;
	i32 err = dx - dy;
	const i32 thickness = thicknessValue;
	while (true) {
		const i32 half = thickness >> 1;
		for (i32 yy = currentY - half; yy < currentY - half + thickness; ++yy) {
			if (yy < 0 || yy >= frameBufferHeight) {
				continue;
			}
			for (i32 xx = currentX - half; xx < currentX - half + thickness; ++xx) {
				if (xx < 0 || xx >= frameBufferWidth) {
					continue;
				}
				const size_t index = static_cast<size_t>(yy) * static_cast<size_t>(frameBufferWidth) + static_cast<size_t>(xx);
				blendFrameBufferPixel(pixels, index, color.r, color.g, color.b, color.a, layer, priority, seq);
			}
		}
		if (currentX == targetX && currentY == targetY) {
			return;
		}
		const i32 e2 = err << 1;
		if (e2 > -dy) {
			err -= dy;
			currentX += sx;
		}
		if (e2 < dx) {
			err += dx;
			currentY += sy;
		}
	}
}

void VdpFrameBufferRasterizer::rasterizeFrameBufferBlit(std::vector<u8>& pixels, u32 frameWidth, u32 frameHeight, const VdpBlitterSource& source, i32 dstX, i32 dstY, i32 dstW, i32 dstH, bool flipH, bool flipV, const VdpFrameBufferColor& color, Layer2D layer, f32 priority, u32 seq) {
	const i32 frameBufferWidth = static_cast<i32>(frameWidth);
	const i32 frameBufferHeight = static_cast<i32>(frameHeight);
	const VdpSurfaceUploadSlot* sourceSlot = m_vdp.resolveFrameBufferExecutionSource(source.surfaceId);
	if (sourceSlot == nullptr) {
		return;
	}
	const auto& sourcePixels = sourceSlot->cpuReadback;
	const size_t sourceStride = static_cast<size_t>(sourceSlot->surfaceWidth) * 4u;
	const i32 sourceWidth = static_cast<i32>(source.width);
	const i32 sourceHeight = static_cast<i32>(source.height);
	i32 srcY = 0;
	i32 srcYRemainder = 0;
	for (i32 y = 0; y < dstH; ++y) {
		const i32 targetY = dstY + y;
		if (targetY >= 0 && targetY < frameBufferHeight) {
			const i32 sampleSourceY = flipV ? sourceHeight - 1 - srcY : srcY;
			i32 srcX = 0;
			i32 srcXRemainder = 0;
			for (i32 x = 0; x < dstW; ++x) {
				const i32 targetX = dstX + x;
				if (targetX >= 0 && targetX < frameBufferWidth) {
					const i32 sampleSourceX = flipH ? sourceWidth - 1 - srcX : srcX;
					const uint32_t sampleX = source.srcX + static_cast<uint32_t>(sampleSourceX);
					const uint32_t sampleY = source.srcY + static_cast<uint32_t>(sampleSourceY);
					if (sampleX < sourceSlot->surfaceWidth && sampleY < sourceSlot->surfaceHeight) {
						const size_t srcIndex = (static_cast<size_t>(sampleY) * sourceStride) + (static_cast<size_t>(sampleX) * 4u);
						const u8 srcA = sourcePixels[srcIndex + 3u];
						if (srcA != 0u) {
							const u8 outA = static_cast<u8>((static_cast<u32>(srcA) * static_cast<u32>(color.a) + 127u) / 255u);
							const u8 outR = static_cast<u8>((static_cast<u32>(sourcePixels[srcIndex + 0u]) * static_cast<u32>(color.r) + 127u) / 255u);
							const u8 outG = static_cast<u8>((static_cast<u32>(sourcePixels[srcIndex + 1u]) * static_cast<u32>(color.g) + 127u) / 255u);
							const u8 outB = static_cast<u8>((static_cast<u32>(sourcePixels[srcIndex + 2u]) * static_cast<u32>(color.b) + 127u) / 255u);
							const size_t dstIndex = static_cast<size_t>(targetY) * static_cast<size_t>(frameBufferWidth) + static_cast<size_t>(targetX);
							blendFrameBufferPixel(pixels, dstIndex, outR, outG, outB, outA, layer, priority, seq);
						}
					}
				}
				srcXRemainder += sourceWidth;
				while (srcXRemainder >= dstW) {
					++srcX;
					srcXRemainder -= dstW;
				}
			}
		}
		srcYRemainder += sourceHeight;
		while (srcYRemainder >= dstH) {
			++srcY;
			srcYRemainder -= dstH;
		}
	}
}

} // namespace bmsx
